Initial APM terminology doc (#133)
elastic/apm: docs/terminology.md

# Terminology

This document describes terms and concepts often used within the APM ecossystem

#### APM
Application Performance Monitoring is the concept of profiling and monitoring systems

#### APM dashboards
Custom Kibana dashboards made for APM. These used to be bundled with Kibana but are now located in the [apm-contrib repo](https://github.com/elastic/apm-contrib/tree/master/apm-ui)

#### APM UI

> The curated UI in Kibana. This is part of the Elastic Basic License.

Reviewer: ```suggestion
The curated UI in Kibana. This is only available with an Elastic Basic License.
```

[Further reading](https://www.elastic.co/guide/en/kibana/7.3/xpack-apm.html)

#### Agent 
An APM agent lives inside an application and will automatically collect APM data (transactions, spans, metrics and errors) and send it to the APM Server.
[Further reading]](https://www.elastic.co/guide/en/apm/get-started/current/components.html#_apm_agents)

#### APM Server
The APM Server receives data from the Elastic APM agents and stores the data into Elasticsearch.
[Further reading](https://www.elastic.co/guide/en/apm/get-started/current/components.html#_apm_server)

#### Instrumentation

> To be able to monitor an application it needs to be _instrumented_. Instrumentation can happen automatically (the Elastic APM agents intrument many things automatically) or manually.

Reviewer: ```suggestion
To be able to monitor an application it needs to be _instrumented_. Instrumentation can happen automatically (the Elastic APM agents instrument many things automatically) or manually.
```

Definitions of automatic and manual are probably useful, perhaps:

automatic: Zero or minimal code changes needed for agent to instrument application
manual: Code changes required, helper libraries are often provided to simplify changes.

Do things like java auto-attach fit here as well?

#### Service

> The application/service being instrumented by APM. A service can be separated by environment.

Reviewer: ```suggestion
The application/service being instrumented by APM. A service is uniquely identified by name + environment.
```

# Terminology

> This document describes terms and concepts often used within the APM ecossystem

Reviewer 2: ```suggestion
This document describes terms and concepts often used within the APM ecosystem.
```

## APM documents

#### Span
Spans contain information about a specific code path that has been executed. They measure from the start to end of an activity, and they can have a parent/child relationship with other spans.
[Further reading](https://www.elastic.co/guide/en/apm/get-started/current/transaction-spans.html)

#### Transaction
Transactions are a special kind of span that have additional attributes associated with them. They describe an event captured by an Elastic APM agent instrumenting a service. You can think of transactions as the highest level of work you’re measuring within a service
[Further reading](https://www.elastic.co/guide/en/apm/get-started/current/transactions.html)

#### Trace
A trace is a grouping of spans and transactions that all share the same `trace.id`

## Sampling

To reduce processing and storage overhead, transactions may be "sampled". Sampling limits the amount of data that is captured for transactions: non-sampled transactions will not record context, and related spans will not be captured.

#### Adaptive sampling
TODO

#### Head based sampling

> Deciding whether to sample a trace before it has started. The decision to sample will often be very simplistic eg. sample 1% of traces.

Reviewer 2: ```suggestion
Deciding whether to sample a trace before it has started. The decision to sample will often be very simplistic e.g. sample 1% of traces.
```

#### Tail-based sampling

> Deciding whether to sample a trace after it has ended. This makes it possible to sample based on _interesting_ events like error occurence, latency etc.

Reviewer 2: ```suggestion
Deciding whether to sample a trace after it has ended. This makes it possible to sample based on _interesting_ events like error occurrence, latency etc.
```

# Terminology

This document describes terms and concepts often used within the APM ecossystem

#### APM

> Application Performance Monitoring is the concept of profiling and monitoring systems

Reviewer 3: ```suggestion
Application Performance Monitoring is the concept of profiling and monitoring services and applications. For instance, it accounts for things like requests per second, but not how much free space is on a disk.
```

#### Distributed Tracing
Distributed tracing is a method for monitoring how applications interact with each other, especially in a microservice architecture.
[Further reading](https://www.elastic.co/guide/en/apm/get-started/current/distributed-tracing.html)

#### Instrumentation

> To be able to monitor an application it needs to be _instrumented_. Instrumentation can happen automatically (the Elastic APM agents instrument many things automatically) or manually.

Reviewer 3: ```suggestion
To be able to monitor an application it needs to be _instrumented_ by an APM Agent. 
Usually not everything is instrumented because that would incur a very large overhead. We are interested only in performance-sensitive activities, like database queries or web requests. Instrumentation can happen automatically (the Elastic APM Agents instrument many frameworks and databases automatically), or manually with a programmatic API.
```
